avocado/core/varianter.py

```diff
@@ -386,51 +326,45 @@ def __init__(self, debug=False):
                in order to provide the right results.
         """
         self.default_params = {}
-        self.variants = None
         self.debug = debug
-        self.data = tree.TreeNodeDebug() if debug else tree.TreeNode()
-        self._mux_path = None
+        self.node_class = tree.TreeNode if not debug else tree.TreeNodeDebug
         self.ignore_new_data = False    # Used to ignore new data when parsed
+        self._variant_plugins = dispatcher.VarianterDispatcher()
+        self._no_variants = None
 
     def parse(self, args):
         """
         Apply options defined on the cmdline
 
         :param args: Parsed cmdline arguments
         """
-        root = copy.deepcopy(self._process_default_params(args))
-        if self.data != tree.TreeNode():
-            # TODO: Do this per-variant-plugin
-            root.merge(self.data)
-            self.variants = MuxTree(root)
-        else:
-            self.variants = False
-        self._mux_path = getattr(args, 'mux_path', None)
-        if self._mux_path is None:
-            self._mux_path = ['/run/*']
+        defaults = self._process_default_params(args)
+        self._variant_plugins.map_method_copy("initialize", args)
+        self._variant_plugins.map_method_copy("update_defaults", defaults)
+        self._no_variants = sum(self._variant_plugins.map_method("__len__"))
 
     def _process_default_params(self, args):
         """
         Process the default params
 
         :param args: Parsed cmdline arguments
         """
-        default_params = tree.TreeNode()
+        default_params = self.node_class()
         for default_param in self.default_params.itervalues():
             default_params.merge(default_param)
         self.default_params = default_params
         # FIXME: Backward compatibility params, to be removed when 36 LTS is
         # discontinued
         if (not getattr(args, "mux_skip_defaults", False) and
                 hasattr(args, "default_avocado_params")):
-            self.default_params.merge(args.default_avocado_params)
+            self.default_params.merge(args.default_avocado_params)  # self.default_params is TreeNode pylint: disable=E1101
```

Comment:
OK I added a commit to address this so we can either keep it, squash it or remove it....